### DISCUSSION TITLE
Getting (and setting) txPower dynamically

### DISCUSSION THREAD

**Discussion by oliv3r** [Koenkk/Z-Stack-firmware#510] [Ideas]
Hey all,

going over the sources, I noticed the following message ID:
```
source/ti/zstack/mt/mt.h:123:#define MT_SYS_SET_TX_POWER                  0x14
```

Which made me wonder, how can I invoke this command? I know I could hack on zigpy-znp; but I want to start locally and small. E.g. 'telnet' to the device and start sending/receiving commands. Obviously I could use picocom and open the serial port; but then i'd need to send the character 14 + some expected value, and then what. 0x14 is not even something that's easily transmittable, as its a non-printable ascii character. So how to do this?

Secondly, I just did this
```diff
@@ -121,7 +121,7 @@ extern "C"
 #define MT_SYS_OSAL_NV_DELETE                0x12
 #define MT_SYS_OSAL_NV_LENGTH                0x13
 #define MT_SYS_SET_TX_POWER                  0x14
-// 0x15 is available for use. Was used for MT_SYS_JAMMER_PARAMETERS
+#define MT_SYS_GET_TX_POWER                  0x15
 #define MT_SYS_SNIFFER_PARAMETERS            0x16
 #define MT_SYS_ZDIAGS_INIT_STATS             0x17
 #define MT_SYS_ZDIAGS_CLEAR_STATS            0x18
@@ -314,6 +314,9 @@ uint8_t MT_SysCommandProcessing(uint8_t *pBuf)
     case MT_SYS_SET_TX_POWER:
       MT_SysSetTxPower(pBuf);
       break;
+    case MT_SYS_GET_TX_POWER:
+      MT_SysGetTxPower();
+      break;
 
 // CC253X MAC Network Processor does not have NV support
 #if !defined( CC253X_MACNP )
@@ -1689,6 +1692,26 @@ static void MT_SysSetTxPower(uint8_t *pBuf)
                                 &status);
 }
 
+/******************************************************************************
+ * @fn      MT_SysGetTxPower
+ *
+ * @brief   Get the transmit power.
+ *
+ * @param   None
+ *
+ * @return  None
+ *****************************************************************************/
+static void MT_SysGetTxPower(void)
+{
+  /* A local variable to hold the signed dBm value of TxPower that is being requested. */
+  int8_t txPower;
+
+  MAP_MAC_MlmeGetReq(MAC_PHY_TRANSMIT_POWER_SIGNED, &txPower);
+
+  // Send back response with the value of the current txPower.
+  MT_BuildAndSendZToolResponse( MT_SRSP_SYS, MT_SYS_SET_TX_POWER, 1, &txPower);
+}
+
 #if defined ( FEATURE_SYSTEM_STATS )
```

where I have no idea if this is even correct or not :p but I want to try and learn. Figured this would make for a cute addition and validate what I am setting is what I am getting.

What I was thinking about, is to see if host programs could adjust the txPower in realtime, like a PID control loop or something. Just playing with the thought for now, but what triggered me was, that after flashing a 20 dBm firmware, it felt like reception was way-worse then when I flashed it at 9 dBm. So to validate my thoughts, I'd have to compile (which takes forever) and flash. Constantly. So being able to tweak the parameters at runtime would be neat.

For now, I don't even know if it compiles, just trying to see if this is something I could do ...

Likewise, I'd be able to use this to see if I can toggle the LED to see if it actually works ;)

**Comment by Koenkk** [comment]
Z2M writes `[254,2,33,15,0,10,38]` when setting transmit power to 10 (second last byte is the tx power)

```
[2024-08-20 21:12:17] debug: 	zh:zstack:znp: SREQ: --> SYS - stackTune - {"operation":0,"value":10}
[2024-08-20 21:12:17] debug: 	zh:zstack:unpi:writer: --> frame [254,2,33,15,0,10,38]
[2024-08-20 21:12:17] debug: 	zh:zstack:unpi:parser: <-- [254,1,97,15,0,111]
[2024-08-20 21:12:17] debug: 	zh:zstack:unpi:parser: --- parseNext [254,1,97,15,0,111]
[2024-08-20 21:12:17] debug: 	zh:zstack:unpi:parser: --> parsed 1 - 3 - 1 - 15 - [0] - 111
[2024-08-20 21:12:17] debug: 	zh:zstack:znp: SRSP: <-- SYS - stackTune - {"value":0}
```

**Reply by oliv3r** [reply]
Ok thanks! I assume that this uses the command `MT_SYS_SET_TX_POWER`, but there's no 'trick' to send bytes other then using high-level tools? Guess I'll have to experiment and mess with 'zigpy' then. I probably also need a second dongle for all this playing, so I won't disturb my current setup too much :D  I quite like the sonoff, but any other idea's for another dongle?

**Reply by Koenkk** [reply]
I really like the Sonoff dongle (also use it in my prod setup atm)

**Reply by oliv3r** [reply]
> I really like the Sonoff dongle (also use it in my prod setup atm)

Same here, but I want a second one to hack on, without taking down my network; so looks like I'll order a second one ;)

While the TI software stack is horrible to work with; it's better then the efr32 bit I suppose, as I'm not seeing any effort to work on that. One could argue it's perfect out of the box though ;)
